overlord/devicestate/devicestate.go

```diff
@@ -1722,72 +1740,125 @@ func CreateRecoverySystem(st *state.State, label string, opts CreateRecoverySyst
 		return nil, err
 	}
 
+	// the task that creates the recovery system doesn't know anything about
+	// validation sets, so we cannot create systems with snaps that are not in
+	// the model.
+	if err := checkForRequiredSnapsNotPresentInModel(model, valsets); err != nil {
+		return nil, err
+	}
+
 	tracker := snap.NewSelfContainedSetPrereqTracker()
 
+	validRevision := func(current snap.Revision, constraints snapasserts.PresenceConstraint) bool {
+		return constraints.Revision.Unset() || current == constraints.Revision
+	}
+
 	var downloadTSS []*state.TaskSet
 	for _, sn := range model.AllSnaps() {
-		rev := revisions[sn.Name]
+		constraints, err := valsets.Presence(sn)
+		if err != nil {
+			return nil, err
+		}
 
-		needsInstall, err := snapNeedsInstall(st, sn.Name, rev)
+		installed, currentRevision, err := installedSnapRevision(st, sn.Name)
 		if err != nil {
 			return nil, err
 		}
 
-		if !needsInstall {
-			info, err := snapstate.CurrentInfo(st, sn.Name)
-			if err != nil {
-				return nil, err
-			}
-			tracker.Add(info)
+		// if the snap is installed, then we must either download it from the
+		// store, have it provided locally, or it must be installed at the
+		// correct revision.
+		//
+		// TODO: in the case that the snap is installed at the wrong revision,
+		// we must provide it either from the store or locally. this is because
+		// doCreateRecoverySystem will install any optional snaps that are
+		// present on the system.
+		required := constraints.Presence == asserts.PresenceRequired || sn.Presence == "required" || installed
```

Comment:
can you remind me why installed also means required? same below for components

Reply:
could you expand the comment to clarify this a bit then? it's a bit strange that is starts with "if ... is installed" while first there are things about constraints